### PR TITLE
[SPARK-26239] File-based secret key loading for SASL.

### PR DESCRIPTION
**Pull request by mccheah**
## What changes were proposed in this pull request?

This proposes an alternative way to load secret keys into a Spark application that is running on Kubernetes. Instead of automatically generating the secret, the secret key can reside in a file that is shared between both the driver and executor containers.

## How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by mccheah** [comment]
@vanzin @vinooganesh @ifilonenko

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 1]
```diff
@@ -367,11 +372,26 @@ private[spark] class SecurityManager(
 
       case _ =>
         require(sparkConf.contains(SPARK_AUTH_SECRET_CONF),
-          s"A secret key must be specified via the $SPARK_AUTH_SECRET_CONF config.")
+          s"A secret key must be specified via the $SPARK_AUTH_SECRET_CONF config")
         return
     }
 
-    secretKey = Utils.createSecret(sparkConf)
+    if (sparkConf.get(AUTH_SECRET_FILE_DRIVER).isDefined
```
Comment (mccheah) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: Also considered adding a validation that file-based auth can only be used for Kubernetes. Omitted it for simplicity, but file-based auth seems dubious for non-K8s contexts.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 2]
```diff
@@ -16,10 +16,13 @@
...
 package org.apache.spark.deploy.k8s.features
 
-import scala.collection.JavaConverters._
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
 
 import io.fabric8.kubernetes.api.model._
 import org.scalatest.BeforeAndAfter
+import scala.collection.JavaConverters._
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala]: ? Hi, @mccheah . We import `java.*` and `scala.*` before any others.

**Comment by tgravescs** [comment]
How is this file protected in kubernetes?  
I actually don't like this idea at least for yarn and other deployments, I see people abusing it (accidentally) and using it in non-secure manner. 
I realize its up to the user to shoot themselves but I would like to keep that to a minimal to where they can shoot themselves.

Do you need updates to the kurbernetes specific docs on how users use this?

**Review by vanzin — COMMENTED** [review]
Mostly nits. But I do agree with Tom that this is more dangerous than helpful outside of k8s. It's too easy to do the wrong thing. So maybe it should be explicitly disallowed there.

**Comment by vanzin** [comment]
(Also, another nit, Spark authentication is not necessarily SASL-based.)

**Comment by mccheah** [comment]
@tgravescs thanks for the feedback!

> How is this file protected in kubernetes?

One can mount secrets into files in Kubernetes, and it's up to Kubernetes to mount these securely. See [here](https://kubernetes.io/docs/concepts/configuration/secret/). We're already allowing support for user-mounted secrets so then it's up to the user to create the secret bytes and point to the right mount location. There are alternative methods, e.g. using Vault. Persistent volume claims that are permisssioned properly can also work.

> I actually don't like this idea at least for yarn and other deployments, I see people abusing it (accidentally) and using it in non-secure manner.

Sure, we can forbid this being used in non-Kubernetes contexts.

> Do you need updates to the kurbernetes specific docs on how users use this?

Docs are updated in the `security.md` file. Specifying how the file is mounted is up to the user. We can suggest using Kubernetes secrets, but again in Kubernetes other mounting options are available.

**Comment by mccheah** [comment]
> How is this file protected in kubernetes?

More precisely it's a core concept in Kubernetes that container boundaries provide strong enough isolation such that data doesn't leak between containers. Outside of fringe cases like `hostPath` volumes and shared persistent volume mounts backed by NFS containers are assumed to have isolated views of disk and mounted volumes.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 3]
```diff
@@ -380,6 +400,12 @@ private[spark] class SecurityManager(
     }
   }
 
+  private def readSecretFile(secretFilePath: String): String = {
+    val secretFile = new File(secretFilePath)
+    require(secretFile.isFile, s"No file found containing the secret key at $secretFilePath.")
+    Base64.getEncoder.encodeToString(Files.readAllBytes(secretFile.toPath))
```
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/scala/org/apache/spark/SecurityManager.scala]: Will check non-empty only

**Comment by mccheah** [comment]
@vanzin 

> (Also, another nit, Spark authentication is not necessarily SASL-based.)

The security doc was updated to reflect this. Thanks!

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala (hunk 2 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala]: Think I fixed this.

File: core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala [hunk 4]
```diff
@@ -440,12 +473,27 @@ class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties {
...
 
+  private def encodeFileAsBase64(secretFile: File) = {
+    Base64.getEncoder.encodeToString(Files.readAllBytes(secretFile.toPath))
+  }
+
+  private def createTempSecretFile(contents: String = "test-secret"): File = {
+    val secretDir = Utils.createTempDir("temp-secrets")
+    val secretFile = new File(secretDir, "temp-secret.txt")
+    Files.write(secretFile.toPath, contents.getBytes(StandardCharsets.UTF_8))
+    secretFile
```
[the remark this answers is not in the document]
Comment (felixcheung) [reply core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala]: ah it's fine

File: core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala [hunk 5]
```diff
@@ -395,9 +397,54 @@ class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties {
     assert(keyFromEnv === new SecurityManager(conf2).getSecretKey())
   }
 
+  test("Use executor-specific secret file configuration.") {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala]: All test names in this class start with lower case, so please follow the pattern.

File: core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala [hunk 6]
```diff
@@ -395,9 +397,54 @@ class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties {
...
+    val conf = new SparkConf()
+      .setMaster("k8s://127.0.0.1")
+      .set(AUTH_SECRET_FILE_DRIVER, Some(secretFileFromDriver.getAbsolutePath))
+      .set(AUTH_SECRET_FILE_EXECUTOR, Some(secretFileFromExecutor.getAbsolutePath))
+      .set(SecurityManager.SPARK_AUTH_CONF, "true")
+    val mgr = new SecurityManager(conf, authSecretFileConf = AUTH_SECRET_FILE_EXECUTOR)
+    assert(encodeFileAsBase64(secretFileFromExecutor) === mgr.getSecretKey())
+  }
+
+  test("Files must be loaded on both the driver and executor and not only one of the two.") {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/SecurityManagerSuite.scala]: "secret file must be defined in both driver and executor". Shorter and clearer.

**Review by vanzin — COMMENTED** [review]
Couple of nits otherwise good to go.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 7]
```diff
@@ -158,6 +162,25 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
...
+      .set("spark.master", "k8s://127.0.0.1")
+    val secMgr = new SecurityManager(conf)
+    secMgr.initializeAuth()
+
+    val step = new BasicExecutorFeatureStep(KubernetesTestConf.createExecutorConf(sparkConf = conf),
+      secMgr)
+
+    val executor = step.configurePod(SparkPod.initialPod())
+    assert(!KubernetesFeaturesTestUtils.containerHasEnvVar(
+      executor.container, SecurityManager.ENV_AUTH_SECRET))
```
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala]: Probably unnecessary, but maybe check contents?

**Review by ifilonenko — APPROVED** [review]
LGTM, just small testing NITthat could be disregarded.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala (hunk 7 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala]: There should be no contents, which is what this assertion is checking.

File: docs/security.md [hunk 8]
```diff
@@ -66,6 +66,50 @@ Kubernetes admin to ensure that Spark authentication is secure.
...
+
+<table class="table">
+<tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
+<tr>
+  <td><code>spark.authenticate.secret.file</code></td>
+  <td>None</td>
+  <td>
+    Path pointing to the secret key to use for securing connections. Ensure that the
+    contents of the file have been securely generated. This file is loaded on both the driver
+    and the executors unless other settings override this (see below).
```
Comment (tgravescs) [inline docs/security.md]: similar add only supported on kubernetes

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 9]
```diff
@@ -419,6 +419,34 @@ package object config {
...
+      .createOptional
+
+  private[spark] val AUTH_SECRET_FILE_DRIVER =
+    ConfigBuilder("spark.authenticate.secret.driver.file")
+      .doc("Path to a file that contains the authentication secret to use. Loaded by the" +
+        " driver. In Kubernetes client mode it is often useful to set a different secret" +
+        " path for the driver vs. the executors, since the driver may not be running in" +
+        " a pod unlike the executors. If this is set, an accompanying secret file must" +
+        " be specified for the executors. The fallback configuration allows the same path to be" +
+        " used for both the driver and the executors when running in cluster mode.")
```
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: The contents of the file must be the same. The locations may be different or the same given the configuration matrix proposed here.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 10]
```diff
@@ -367,11 +371,18 @@ private[spark] class SecurityManager(
 
       case _ =>
         require(sparkConf.contains(SPARK_AUTH_SECRET_CONF),
-          s"A secret key must be specified via the $SPARK_AUTH_SECRET_CONF config.")
+          s"A secret key must be specified via the $SPARK_AUTH_SECRET_CONF config")
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: Undo this change.

**Review by vanzin — COMMENTED** [review]
Merging to master. I'll fix the comment during merge.

**Comment by mccheah** [comment]
Oh I fixed it just now
